Create a UUID from given LG soundbar device name

chrismccurdy opened the pull request:
* If soundbar API provides no UUID in the responses, one is created based on the name the user gives the device.
* Check for dictionary keys before adding to queue & log if missing.
* Check for existence of dictionary keys before attempting to address them.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Fixes home-assistant#77524 by creating a UUID from the soundbar device's user-provided name when the soundbar API doesn't provide a UUID.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #77524
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

chrismccurdy committed:
Create a UUID from given device name (home-assistant#77524)

* If soundbar API provides no UUID in the responses, one is created
  based on the name the user gives the device.
* Check for dictionary keys before adding to queue & log if missing.
* Check for existence of dictionary keys before attempting to address
  them.

MartinHjelmare commented on homeassistant/components/lg_soundbar/config_flow.py:
```diff
@@ -1,51 +1,75 @@
+                "homeassistant.components.lg_soundbar."
+                + sb_name.replace(" ", "_").lower()
+            )
+            sb_uuid = uuid.uuid5(uuid.NAMESPACE_OID, oid)
```
I don't think this is within our unique_id requirements. The unique_id should be stable and not be able to change by user input.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-id-requirements

chrismccurdy replied:
What's the alternative since it looks like this type of device isn't providing any information that could be used as a UUID according to those requirements? Should it follow the `_async_handle_discovery_without_unique_id` suggestion above the requirements on that document?

MartinHjelmare replied:
Yeah, I think that's the way to solve this.

MartinHjelmare replied:
Although, there's no discovery here.

MartinHjelmare replied:
If there's no uuid, I'd abort if there's already a configured config entry, otherwise we set the integration domain as unique_id.

chrismccurdy replied:
I don't quite understand. Isn't the integration domain required to be equal to the directory of the component? https://developers.home-assistant.io/docs/creating_integration_manifest#domain

MartinHjelmare replied:
Yes?

chrismccurdy replied:
I think I'm misunderstanding this statement:

> If there's no uuid, I'd abort if there's already a configured config entry, otherwise we set the integration domain as unique_id.

I don't understand what "set the integration domain as unique_id" means. The integration domain is "lg_soundbar" for this component, or am I mixing up concepts?

MartinHjelmare replied:
I mean that we can set the `unique_id` to the domain of the integration.

chrismccurdy replied:
Thanks, I've got it now. I've pushed a change for that to my branch.

emontnemery replied:
If the config entry's unique id is set to `lg_soundbar` as a fallback, it will only be possible to set up a single device from which we can't read an identifier. Isn't that an unwanted limitation?

Do you have a log from <https://github.com/home-assistant/core/issues/77524> which shows what the device sends back as a response for the requests to get the UUID?

emontnemery replied:
As suggested by @MartinHjelmare on Discord:
> We can also not set a unique id for the case when the UUID can't be read and abort on same host and use the config entry id as unique id for the entities

I think this is a better solution, it will allow multiple sound bars which don't support reading UUID 👍

MartinHjelmare replied:
An alternative is to not set a unique id for the config entry if missing an uuid and just abort if trying to configure the same host, and use the config entry id as unique_id for the entities in that case.

We have this helper to check for same data among existing config entries of the integration and abort:
https://github.com/home-assistant/core/blob/727dcd6df6e240ec74084ffdb0baea22e3946745/homeassistant/config_entries.py#L1352-L1356

emontnemery replied:
I think that's better solution @MartinHjelmare 👍

chrismccurdy replied:
I've modified it to only assign a unique_id if the device returns a UUID and to abort if an integration configuration with the same host and port already exists.

https://github.com/chrismccurdy/home-assistant/blob/2d8374c0b3c2ba6bf5c465c8fd08a75a86362072/homeassistant/components/lg_soundbar/config_flow.py#L94-L99

MartinHjelmare replied:
Good! We may want to give the entities a unique_id in the case that the config entry doesn't have a unique_id. We can use the config entry `entry_id` as entity unique_id in that case.

chrismccurdy replied:
How do we get the `entry_id` in the config flow? It appears that the `entry_id` isn't generated until ConfigEntriesFlowManager.async_finish_flow is called and the FlowResult returned by async_step_user is added to the config_entries.

MartinHjelmare replied:
I'm talking about the `Entity` instances of the media_player platform. When setting up the media player the config entry will have been created. The `entry_id` is an attribute of the `ConfigEntry` instance.

chrismccurdy committed:
Changed to abide by UUID rules

MartinHjelmare commented on homeassistant/components/lg_soundbar/config_flow.py:
```diff
@@ -70,13 +86,18 @@ async def async_step_user(self, user_input=None):
-            }
-            return self.async_create_entry(title=details["name"], data=info)
+            if len(details) != 0:
+                unique_id = DOMAIN
```
We should only allow this if there are no existing config entries for the integration.

chrismccurdy committed:
Abort if no UUID and host already configured

chrismccurdy committed:
Set media player entity unique_id

If ConfigEntry.unique_id is not set, use the ConfigEntry.entity_id for
the LGDevice's unique_id

MartinHjelmare reviewed (commented):
What's the test coverage now for the config flow module?

chrismccurdy commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -209,6 +213,214 @@ def temescal_side_effect(addr, port, callback):
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.lg_soundbar.config_flow.temescal", return_value=Mock()
```
```---------- coverage: platform linux, python 3.9.15-final-0 -----------
Name                                                  Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------
homeassistant/components/lg_soundbar/__init__.py         19      9    53%   20-30, 37-38
homeassistant/components/lg_soundbar/config_flow.py      72      4    94%   65-68
homeassistant/components/lg_soundbar/const.py             2      0   100%
-----------------------------------------------------------------------------------
TOTAL                                                    93     13    86%

chrismccurdy replied:
Test coverage is now 100% for the config flow module.

```---------- coverage: platform linux, python 3.9.15-final-0 -----------
Name                                                  Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------
homeassistant/components/lg_soundbar/__init__.py         19      9    53%   20-30, 37-38
homeassistant/components/lg_soundbar/config_flow.py      70      0   100%
homeassistant/components/lg_soundbar/const.py             2      0   100%
-----------------------------------------------------------------------------------
TOTAL                                                    91      9    90%

chrismccurdy committed:
Cleaned up return for test_connect

* test_connect will return the details dict outside the try-block with
  whatever values were populated
* tests cleaned up to reduce redundancy

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -209,6 +208,208 @@ def temescal_side_effect(addr, port, callback):
+        )
+        await hass.async_block_till_done()
+
+    print("result2 == [%s]", result2)
```
```suggestion
```

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -209,6 +208,208 @@ def temescal_side_effect(addr, port, callback):
+        )
+        await hass.async_block_till_done()
+
+    print("result2 " + str(result2))
```
```suggestion
```

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -209,6 +208,208 @@ def temescal_side_effect(addr, port, callback):
+    ) as mock_temescal, patch(
+        "homeassistant.components.lg_soundbar.config_flow.Queue"
+    ) as mock_q, patch(
+        "homeassistant.components.lg_soundbar.config_flow.LGSoundbarConfigFlow.async_set_unique_id"
```
Don't patch a core helper of the flow. Assert the config entry unique_id attribute instead. The config entry instance is available under the `"result"` key in the flow result dict for the create entry result.

MartinHjelmare replied:
We don't seem to assert the unique_id attribute now. Did I miss it?

chrismccurdy replied:
unique_id attribute is being asserted at https://github.com/chrismccurdy/home-assistant/blob/11cacaa6f543029929ff097ed8699afba765a561/tests/components/lg_soundbar/test_config_flow.py#L284-L286

chrismccurdy replied:
unique_id is now being checked against the `result` key in the flow result dict. Also added assertions for the other tests to verify that the case where the unique_id is set to the device uuid.

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -209,6 +208,208 @@ def temescal_side_effect(addr, port, callback):
+    assert result["errors"] == {}
+
+    with patch("homeassistant.components.lg_soundbar.config_flow.temescal"), patch(
+        "homeassistant.components.lg_soundbar.config_flow.test_connect",
```
Don't patch the integration code in the config flow. We want to aim at patching the library and as few parts of our code in the config flow as possible.

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -209,6 +208,208 @@ def temescal_side_effect(addr, port, callback):
+        mock_name_q.get.side_effect = Empty()
+
+        tmock = mock_temescal.temescal
+        tmock.return_value = Mock()
```
```suggestion
```

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -209,6 +208,208 @@ def temescal_side_effect(addr, port, callback):
+        mock_name_q.get.return_value = "name"
+
+        tmock = mock_temescal.temescal
+        tmock.return_value = Mock()
```
```suggestion
```

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -209,6 +208,208 @@ def temescal_side_effect(addr, port, callback):
+        "homeassistant.components.lg_soundbar.async_setup_entry", return_value=True
+    ) as mock_setup_entry:
+        tmock = mock_temescal.temescal
+        tmock.return_value = Mock()
```
```suggestion
```

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -209,6 +208,208 @@ def temescal_side_effect(addr, port, callback):
+    ) as mock_setup_entry:
+        mock_q.side_effect = [mock_uuid_q, mock_name_q]
+        mock_uuid_q.empty.return_value = True
+        mock_uuid_q.get.side_effect = Empty
```
Instead of patching the queue, I'd patch the timeout of the queue to 0. This is easy if we define the timeout in a module level constant in the config flow. If needed we can sleep a bit in the callback to make sure the timeout always is faster than the callback.

chrismccurdy committed:
Fixed and removed test redundancies

chrismccurdy committed:
Added optional delay to mocked callback

MartinHjelmare commented on homeassistant/components/lg_soundbar/config_flow.py:
```diff
@@ -18,6 +18,13 @@
+QUEUE_TIMEOUT = 10
+
+
+def queue_timeout() -> float:
```
We don't need this. Just use the constant directly.

MartinHjelmare commented on homeassistant/components/lg_soundbar/config_flow.py:
```diff
@@ -54,8 +61,8 @@ def msg_callback(response):
             connection.get_product_info()
-        details["name"] = name_q.get(timeout=10)
-        details["uuid"] = uuid_q.get(timeout=10)
+        details["name"] = name_q.get(timeout=queue_timeout())
+        details["uuid"] = uuid_q.get(timeout=queue_timeout())
```
```suggestion
        details["name"] = name_q.get(timeout=QUEUE_TIMEOUT)
        details["uuid"] = uuid_q.get(timeout=QUEUE_TIMEOUT)
```

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -97,35 +120,20 @@ async def test_form_uuid_present_in_both_functions_uuid_q_empty(hass):
-
     with patch(
+        "homeassistant.components.lg_soundbar.config_flow.queue_timeout",
+        return_value=0.1,
```
```suggestion
        "homeassistant.components.lg_soundbar.config_flow.QUEUE_TIMEOUT",
        new=0.1,
```

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -159,34 +164,20 @@ async def test_form_uuid_present_in_both_functions_uuid_q_not_empty(hass):
-    mock_name_q = MagicMock()
-
     with patch(
+        "homeassistant.components.lg_soundbar.config_flow.queue_timeout",
+        return_value=0.1,
```
```suggestion
        "homeassistant.components.lg_soundbar.config_flow.QUEUE_TIMEOUT",
        new=0.1,
```

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -263,41 +242,19 @@ async def test_form_uuid_not_provided_by_api(hass):
-    mock_name_q = MagicMock()
-
     with patch(
+        "homeassistant.components.lg_soundbar.config_flow.queue_timeout",
+        return_value=0.1,
```
```suggestion
        "homeassistant.components.lg_soundbar.config_flow.QUEUE_TIMEOUT",
        new=0.1,
```

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -395,10 +326,13 @@ async def test_no_uuid_host_already_configured(hass):
-        return_value={"name": "name"},
-    ):
+    with patch(
+        "homeassistant.components.lg_soundbar.config_flow.queue_timeout",
+        return_value=0.1,
```
```suggestion
        "homeassistant.components.lg_soundbar.config_flow.QUEUE_TIMEOUT",
        new=0.1,
```

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -22,13 +26,21 @@ def create_temescal_response(msg: str, data: Optional[dict] = None):
+                sleep(delay)
+                callback(response)
+
+            thread = threading.Thread(target=run_thread, daemon=True)
```
I think we can use `hass.add_job` to schedule the whole `invoke_callback` to the thread pool. Then we don't need a specific thread here.

MartinHjelmare commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -99,35 +132,18 @@ async def test_form_uuid_present_in_both_functions_uuid_q_empty(hass):
+            mac_info_dev={"s_uuid": "uuid"},
+            product_info={"s_uuid": "uuid"},
+            info={"s_user_name": "name"},
+            delay=2.0,
```
Why do we add a delay here? This is not the test that should timeout in the queue. We don't want to have slow tests.

chrismccurdy replied:
This is to test a delay in the callback causing `uuid_q.empty()` at config_flow.py:62 to return True, which may then cause `attr_q.put_nowait(data)` on line 46 to raise Full since the queue maxsize is 1. Without the delay in the test, the item is added to the queue immediately and the check if it's empty always returns False for this test.

chrismccurdy replied:
Changing the callback invocation to use `hass.add_job` instead of a new thread looks like it introduces enough of a delay to enable this test to take the expected path without additional workarounds. Thanks for the tip.

chrismccurdy committed:
lg_soundbar test improvements

* changed the method of mocking the queue timeout period
* replaced threading with hass.add_job for running temescal callback
  asynchronously

chrismccurdy commented on tests/components/lg_soundbar/test_config_flow.py:
```diff
@@ -8,6 +9,46 @@
+        return response
+
+    def temescal_side_effect(addr, port, callback):
+        async def async_callback(response):
```
I tried removing the intermediate callback, but that prevents config_flow lines 42-43 from getting covered by that test case.

MartinHjelmare replied:
I fixed it. :+1:

chrismccurdy committed:
changed method of verifying if unique_id was set

chrismccurdy committed:
Merge branch 'dev' into lg_soundbar_77524

MartinHjelmare committed:
Clean up tests

MartinHjelmare committed:
Ignore init module coverage

MartinHjelmare reviewed (approved):
Thanks!

MartinHjelmare commented:
Codecov is confused. This is the coverage now:

```
---------- coverage: platform linux, python 3.9.15-final-0 -----------
Name                                                  Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------
homeassistant/components/lg_soundbar/config_flow.py      70      0   100%
homeassistant/components/lg_soundbar/const.py             2      0   100%
-----------------------------------------------------------------------------------
TOTAL                                                    72      0   100%

```

chrismccurdy commented:
Thanks for your help, @MartinHjelmare.
